### ISSUE TITLE
conda installation

### ISSUE THREAD

**Issue by sebastian-luna-valero** [libvips/libvips#673]
Hello,

I have seen this conda package for libvips:
https://anaconda.org/zegami/libvips

However, it has some limitations (e.g. openslide is not included).

Is there anyone trying to build a conda package for libvips, and make it available either through [conda-forge](https://conda-forge.github.io/) or [bioconda](https://bioconda.github.io/)?

Best regards,
Sebastian

**Comment by sebastian-luna-valero** [comment]
@eled2 and I are planning to update the libvips recipe for conda, and include support for openslide (which is already in [bioconda](https://github.com/bioconda/bioconda-recipes/tree/master/recipes/openslide)). As a starting point, we'll be using this recipe:

https://github.com/zegami/libvips-conda-recipe

We would be grateful if someone could provide support with compilation/installation woes during our journey.

**Comment by strets123** [comment]
@sebastian-luna-valero I have provided a further changeset to the libvips conda package and added travis ci and a dockerfile. You can see the repo here:

https://github.com/strets123/libvips-conda-recipe

A few helpful notes:

If you are trying to move to python 3.6 you may need to re build various of my packages on zegami channel. If the recipes are not available they can be found inside the tarballs of the downloaded conda packages.

Looking at the available openslide conda builds I am not sure if you need the pyopenslide, if so you will have to rebuild it using python 3. If not it should just be a case of adding the bioconda to the condarc in this recipe.

**Comment by jcupitt** [comment]
Hello, it looks good, I'd say:

* libvips calls the C openslide API and does not expose any openslide itself, so pyopenslide is not required
* I would add `libgsf`, it's required for deepzoom/zoomify/google maps save
* `libexif` is required for JPEG and WebP metadata handling, I'd enable that
* `fftw3` is used for FFT transforms, it should be harmless to enable
* Are `libffi` and `bzip2` required? I can't think of a way they get used, but perhaps they are an indirect dependency. Do `meta.yaml` files have to list indirect dependencies as well?
* Could you change `home:` to be https://jcupitt.github.io/libvips ? That's the new homepage, the soton one is going away soon
* libvips installs for python2 or python3 depending on the setting of `PYTHON`. If you need to support both, you may need to run the install twice, or do some extra `cp`s after install

**Comment by jcupitt** [comment]
Oh, I guess you could add imagemagick or graphicsmagick too. They are useful for things like DICOM image load, which is important for medical imaging with libvips (what I do with it). I've fixed up the DICOM handling in imagemagick and it works quite well now, so I'd suggest that. IM7 is very slow, so IM6 is probably the best one. 

You could also enable `libmatio` which adds Matlab load, another useful feature for scientific work with libvips.

**Comment by jcupitt** [comment]
Sorry, one more thing, libvips comes with a test suite written in Python. It only takes a couple of minutes to run, so it would be worth enabling (in my opinion). Just add a `make check`. It works with py2 and py3.

**Comment by strets123** [comment]
@jcupitt the dependencies removed were an attempt to fully comply with the LGPL license which I think ended up being impossible. @sebastian-luna-valero you can add all of these back in. Normally you would not need to list indirect dependencies but I think Cairo needs this and it was not in the cairo recipe so I put it in to save time.

As far as python 2 and python 3 are concerned yes they can all be supported. Making this work in the first place required a fair amount of fishing around conda for the dependencies some of which don't have working python 2 builds or working python 3.6 builds for that matter.

**Comment by sebastian-luna-valero** [comment]
Thanks both!

Following the most recent instructions in , I have this initial recipe:

* build.sh
```
#!/usr/bin/env bash

./autogen.sh
make
make install
```

* meta.yml
```
{% set name = "libvips" %}
{% set version = "8.5.6" %}
{% set sha256 = "19a77ab200eb0ddfcd8babab130fe43c7730880d1f1fdfbdd8def519af32c4b8" %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  fn: vips-{{ version }}.tar.gz
  url: https://github.com/jcupitt/libvips/releases/download/v{{ version }}/vips-{{ version }}.tar.gz
  sha256: {{ sha256 }}

build:
  number: 0
  skip: true  # [win or osx]

requirements:
  build:
    - automake 1.15.*
    - make 4.2.*
    - cfitsio 3.410.*
    - fftw 3.3.*
    - giflib 5.1.*
    - gobject-introspection 1.51.*
    - graphicsmagick 1.3.*
    - lcms 1.19.*
    - libjpeg-turbo 1.5.*
    - libmatio 1.5.*
    - libpng 1.6.*
    - libtiff 4.0.*
    - libwebp 0.5.*
    - openslide 3.4.*
    - pkg-config 0.29.*
    - poppler 0.52.*
    - swig 3.0.*
#    - gtk-doc # http://jcupitt.github.io/libvips/install.html
#    - imagemagick # https://github.com/jcupitt/libvips/issues/673
#    - libexif # https://github.com/jcupitt/libvips/issues/673
#    - libgsf  # https://github.com/jcupitt/libvips/issues/673
#    - librsvg # http://jcupitt.github.io/libvips/install.html
#    - openexr  # http://jcupitt.github.io/libvips/install.html
#    - orc      # http://jcupitt.github.io/libvips/install.html
#    - pangoft2 # http://jcupitt.github.io/libvips/install.html
  run:
    - cfitsio 3.410.*
    - fftw 3.3.*
    - giflib 5.1.*
    - gobject-introspection 1.51.*
    - graphicsmagick 1.3.*
    - lcms 1.19.*
    - libjpeg-turbo 1.5.*
    - libmatio 1.5.*
    - libpng 1.6.*
    - libtiff 4.0.*
    - libwebp 0.5.*
    - openslide 3.4.*
    - pkg-config 0.29.*
    - poppler 0.52.*
    - swig 3.0.*
#    - gtk-doc # http://jcupitt.github.io/libvips/install.html
#    - imagemagick # https://github.com/jcupitt/libvips/issues/673
#    - libexif # https://github.com/jcupitt/libvips/issues/673
#    - libgsf  # https://github.com/jcupitt/libvips/issues/673
#    - librsvg # http://jcupitt.github.io/libvips/install.html
#    - openexr  # http://jcupitt.github.io/libvips/install.html
#    - orc      # http://jcupitt.github.io/libvips/install.html
#    - pangoft2 # http://jcupitt.github.io/libvips/install.html

test:
  requires:
    - make 4.2.*

  commands:
    - cd $SRC_DIR
#    - export C_INCLUDE_PATH=$PREFIX/include
    - make check

about:
  home: https://jcupitt.github.io/libvips
  license: LGPL-2.1
  license_family: LGPL
  license_file: COPYING
  summary: 'A fast image processing library with low memory needs'

  description: |
    libvips is a demand-driven, horizontally threaded image processing
    library. Compared to similar libraries, libvips runs quickly and uses
    little memory.

    It has around 300 operations covering arithmetic, histograms, convolution,
    morphological operations, frequency filtering, colour, resampling, statistics
    and others. It supports a large range of numeric formats, from 8-bit int to
    128-bit complex. Images can have any number of bands. It supports a good range
    of image formats, including JPEG, TIFF, OME-TIFF, PNG, WebP, FITS, Matlab,
    OpenEXR, PDF, SVG, HDR, PPM, CSV, GIF, Analyze, DeepZoom, and OpenSlide. It can
    also load images via ImageMagick or GraphicsMagick, letting it load formats like DICOM.

    It comes with bindings for C, C++, Python and the command-line. Full bindings are
    available for Ruby, PHP, Go, and Lua. libvips is used as an image processing engine
    by sharp (on node.js), bimg, sharp for Go, carrierwave-vips, mediawiki, PhotoFlow
    and others. The official libvips GUI is nip2, a strange combination of a spreadsheet
    and an photo editor.

  doc_url: https://jcupitt.github.io/libvips/API/current
  dev_url: https://github.com/jcupitt/libvips

extra:
  recipe-maintainers:
    - sebastian-luna-valero
    - strets123
```

Please note that in the `meta.yml` I left commented out those dependencies that are not available yet in conda. Also, I would like to get the recipe working for Linux first, and after that we could think about instructions for Mac and Windows.

My questions are:
* @strets123 do you want to be listed as maintainer in this recipe?
* I am trying to use the `autogen.sh` script but `gtk-doc` is not available as a conda package. @jcupitt what should be easier, adding `gtk-doc` to conda or customising the `build.sh` script with proper `configure` and `make` statements as @strets123 's recipe [here](https://github.com/strets123/libvips-conda-recipe/blob/master/libvips/build.sh) ?

**Comment by jcupitt** [comment]
Hello again, don't use `autogen`, just run `configure`. You'll be able to remove quite a few useless dependencies. 

The build process works like this:

1. The git repository just contains `autogen.sh`. If you want to build from git, you have to run this. It uses your locally installed autotools, plus gtk-doc and swig, to generate all the platform-independent build files. 
2. Once `configure` is made, you can run `make dist`. This makes a release tarball containing all of the source files, plus all of the generated platform independent files. This tarball is what is uploaded to the `release` area on github (the thing you reference with `https://github.com/jcupitt/libvips/releases/download/v{{ version }}/vips-{{ version }}`).
3. From the release tarball, you can just run `configure`. This is a highly-portable shell script and contains as few dependencies as possible. This in turn makes a build system for the local machine, which you then run with `make`.
4. Confusingly, the tarball also contains the `autogen.sh` file, but that's only there because it's good practice to include all sources. It's used earlier in the process, you don't use it for the final configure. 

Therefore: just run `configure`, remove `gtk-doc`, remove `swig`, remove `automake`.

**Comment by jcupitt** [comment]
Other comments: 

* libvips can use either graphicsmagick or imagemagick, but not both at the same time. You've picked GM, so I'd remove IM.
* It's a shame that Orc and libexif do not have anaconda packages. Orc gives a very nice speed-boost, perhaps 4x or 5x faster for some operations, and libexif is essential (obviously) for exif handling. libgsf is required for deepzoom pyramid write.
* I would configure with `-O3`. libvips is marked up for vectorisation, and O3 is the gcc optimisation level that enables the vectoriser.

**Comment by sebastian-luna-valero** [comment]
Great, many thanks!

What's the best way to test that the package is working properly apart from running `make test`? I would be particularly interested in checking that the `openslide` dependency is running fine.

**Comment by jcupitt** [comment]
Just run `make check`. The openslide loader is tested as part of the test suite.

You can run individual parts of it like this:

```
$ cd vips-x.y.x/test
$ ./test_foreign.py 
...................
----------------------------------------------------------------------
Ran 19 tests in 8.455s

OK
```

`test_foreign` tests all the foreign load functions, which includes openslide. Or to test py3:

```
$ python3 test_foreign.py 
...................
----------------------------------------------------------------------
Ran 19 tests in 7.664s

OK
```

**Comment by sebastian-luna-valero** [comment]
Thanks!

I saw the `test_all.py` script, and thought it would be best to use it but here is what I get:
```
python test_all.py
Traceback (most recent call last):
  File "test_all.py", line 8, in <module>
    from test_arithmetic import *
  File "/conda/sebastian/py3/conda-bld/libvips_1497892514588/work/vips-8.5.6/test/test_arithmetic.py", line 11, in <module> 
    gi.require_version('Vips', '8.0')
AttributeError: module 'gi' has no attribute 'require_version'
```

Could you please help?

**Comment by jcupitt** [comment]
Do you have python-gobject installed? 

On Ubuntu I have `/usr/lib/python3/dist-packages/gi` full of stuff. It can be in site-packages too, I'm not sure how Anaconda lays things out.

**Comment by sebastian-luna-valero** [comment]
[python-gobject](https://pygobject.readthedocs.io) is not available in conda, I'm afraid.

Do we really need `python-gobject` to test the installation? or would `make check` suffice? It looks to me like a complex tool to add to conda, so not sure whether it's worth it right now.

**Comment by sebastian-luna-valero** [comment]
By the way, I am running `vips -l` and seems to work fine. Do you suggest any other similar commands that I could add to test the package after it has been built?

**Comment by jcupitt** [comment]
Yes, the vips Python binding is built on python-gobject, it won't work without that. python-gobject isn't that large, it's just used as a base for a lot of other stuff (gtk+, gstreamer, gio, libvips, py-cairo etc. etc.)

The structure is:

1. C funcs in libvips implement image processing stuff.
2. libvips uses GObject as an object framework, and that includes a lot of stuff for introspection.
3. libvips also has a set of special comments in the headers. 
4. `gobject-introspection` uses a combination of static analysis of the headers, plus dynamic analysis of the running library, to generate a .gir file (a huge lump of XML)
5. The .gir file is processed to produce the API docs http://jcupitt.github.io/libvips/API/current/
6. The .gir file is also processed to produce a .typelib. This is a much smaller binary file containing just the type signatures of the API and information on the library class structure. 
7. Using python-gobject, you can load a .typelib file and directly call into any gobject-based library.
8. The API you get is rather ugly, so libvips has a set of overrides (written in Python) which try to make a nice Pythonesque interface (this is the `Vips.py` file that gets installed).

**Comment by jcupitt** [comment]
Have a look in `test/`, there are a set of shell-scripts (mostly run by `make check`) which you can try. For example:

```
john@kiwi:~/GIT/libvips/test (master)$ ./test_formats.sh 
testing йцук.jpg v ... ok
testing йцук.jpg tif ... ok
testing йцук.jpg tif[compression=jpeg] ... ok
testing йцук.jpg tif[compression=deflate] ... ok
testing йцук.jpg tif[compression=packbits] ... ok
testing йцук.jpg tif[compression=jpeg,tile] ... ok
testing йцук.jpg tif[compression=jpeg,tile,pyramid] ... ok
testing йцук.jpg png ... ok
testing йцук.jpg jpg ... ok
testing йцук.jpg webp ... ok
testing йцук.jpg ppm ... ok
testing йцук.jpg pfm ... ok
testing йцук.jpg fits ... ok
testing mono.v csv ... ok
testing cmyk.v jpg ... ok
testing cmyk.v tif ... ok
testing cmyk.v tif[compression=jpeg] ... ok
testing cmyk.v tif[compression=jpeg,tile] ... ok
testing cmyk.v tif[compression=jpeg,tile,pyramid] ... ok
testing rad.v hdr ... ok
testing mono.v raw ... ok
testing йцук.jpg raw ... ok
testing blankpage.pdf pdfload ... ok
testing blankpage.svg svgload ... ok
testing trans-x.gif gifload ... ok
testing sample.mat matlab ... ok
testing dzsave йцук.jpg .zip ... ok
testing copy йцук.jpg .dz ... ok
testing copy йцук.jpg .dz[container=zip] ... ok
```

**Comment by strets123** [comment]
Hi Sebastian, python gobject is available, it 

anaconda search -t conda pygobject
    Using Anaconda API: https://api.anaconda.org
    Run 'anaconda show <USER/PACKAGE>' to get more details:
    Packages:
         Name                      |  Version | Package Types   | Platforms      
         ------------------------- |   ------ | --------------- | ---------------
         ???/pygobject3            |   3.12.0 | conda           | linux-64       
         KristanArmstrong/pygobject |   3.14.0 | conda           | win-64         
                                              : PyGObject is a Python extension module that gives clean and consistent access to the entire GNOME software platform through the use of GObject Introspection.
         fallen/pygobject          |   3.14.0 | conda           | linux-64, linux-32
                                              : PyGObject is a Python extension module that gives clean and consistent access to the entire GNOME software platform through the use of GObject Introspection.
         jeanconn/pygobject        |          | conda           | linux-64       
         jithinpr2/pygobject       |   3.14.0 | conda           | win-64         
                                              : PyGObject is a Python extension module that gives clean and consistent access to the entire GNOME software platform through the use of GObject Introspection.
         jithinpr2/pygobject3      |   3.18.2 | conda           | win-64         
                                              : Access to GObject-based libraries in Python.
         jochym/pygobject2         |   2.28.6 | conda           | linux-64       
         klauer/pygobject3         |   3.12.0 | conda           | linux-64       
                                              : Access to GObject-based libraries in Python.
         licode/pygobject3         |   3.12.0 | conda           | linux-64       
         lightsource2-tag/pygobject3 |   3.12.0 | conda           | linux-64       
         lightsource2-testing/pygobject3 |   3.12.0 | conda           | linux-64       
         lightsource2/pygobject3   |   3.12.0 | conda           | linux-64       
                                              : Access to GObject-based libraries in Python.
         mdryden/pygobject3        |   3.24.2 | conda           | osx-64         
                                              : Access to GObject-based libraries in Python.
         ostrokach/pygobject       |   3.18.2 | conda           | linux-64       
                                              : PyGObject is a Python extension module that gives clean and consistent access to the entire GNOME software platform through the use of GObject Introspection.
         pkgw-forge/pygobject3     |   3.18.2 | conda           | linux-64, osx-64
         pkgw/pygobject3           |          | conda           | linux-64, osx-64
         ska/pygobject             |   2.20.0 | conda           | linux-64       
                                              : Pipeline running tools
         vgauthier/pygobject       |   3.10.0 | conda           | osx-64         
                                              : Python Bindings for GLib/GObject/GIO/GTK+
         zegami/pygobject          |   3.14.0 | conda           | linux-64       
         zegami/pygobject3         |   3.12.0 | conda           | linux-64       
                                              : Access to GObject-based libraries in Python.

**Comment by sebastian-luna-valero** [comment]
Thanks @strets123 !

However, I am trying to use either [conda-forge](https://conda-forge.github.io/) or [bioconda](https://bioconda.github.io/) as these conda channels follow best practices and a reference build environment using containers, which should generate highly portable conda packages.

I am now in conversations with a [PyGObject](https://git.gnome.org/browse/pygobject/) developer to make it happen. Will report back either today or tomorrow.

**Comment by sebastian-luna-valero** [comment]
Hi All,

Sorry for the delay. I am uploading `pygobject` to `conda-forge`:
https://github.com/conda-forge/staged-recipes/pull/3168

After that, the plan is to upload `libvips` to `bioconda`:
https://github.com/bioconda/bioconda-utils/issues/132#issuecomment-309807210

Best regards,
Sebastian

**Comment by strets123** [comment]
Great work, thanks Sebastian. I am happy to be listed as a maintainer for the libvips recipe if that is helpful,

**Comment by sebastian-luna-valero** [comment]
Thanks @strets123, will do.

`pygobject` has now been added to `conda-forge`:
https://github.com/conda-forge/pygobject-feedstock

Now, `make check` fails when building `libvips`. I have attached a copy of `test/test-suite.log`
[output.txt](https://github.com/jcupitt/libvips/files/1101494/output.txt). @jcupitt could you please have a look and let me know how to proceed?

**Comment by jcupitt** [comment]
Hey, almost there!

It seems to fail if pangoft2 or libexif (at a guess) are missing. I'll see if I can make the test suite pass without those libraries.

**Comment by sebastian-luna-valero** [comment]
Brilliant, many thanks @jcupitt !

**Comment by jcupitt** [comment]
Ooof, sorry for the delay. The head of 8.5 should pass if text is missing. I'm not sure what to do about the other failures, they look like a char encoding issue. But I don't know why anaconda should fail on this test, but plain Python passes.

Perhaps anaconda Python does not know about the encoding comments at the top? Or does it default to ASCII for source files?

I've become fed up with the problems gobject-introspection in causing. There's a new Python binding here which just uses cffi:

https://github.com/jcupitt/pyvips

It works on Python 2.7 - 3.6, on Windows and Linux (not tried macOS yet, but it should be easy), with CPython and PyPy, and installs as a simple pip package, just needing brew or whatever to generate a shared library. I'm just doing the docs now.

**Comment by sebastian-luna-valero** [comment]
Thanks, John.

I am on leave until September, and will have a look back then!

**Comment by vishnuprasadh** [comment]
@jcupitt let me know if I can help test on Mac. Iam on Mac. Thanks to u and @sebastian for getting it this far. Iam really eager to get going with this for my project :-=

**Comment by sebastian-luna-valero** [comment]
Hello,

Many thanks for your patient. I have opened a pull request in bioconda:
https://github.com/bioconda/bioconda-recipes/pull/5894

where we all can see progress and comment on it.

Please let me know your thoughts.

**Comment by MatthiasKohl** [comment]
Hi,
I'm interested in having a bioconda recipe for libvips as well. Unfortunately, it seems like nothing happened since bioconda updated their build system to Circle CI (https://github.com/bioconda/bioconda-recipes/pull/5894#issuecomment-360798835) and I'm not very knowledgeable in this regard.
Is there something I can do to help you progress with this ?
Thanks!

**Comment by sebastian-luna-valero** [comment]
Thanks @MatthiasKohl 

I was waiting for @jcupitt 's feedback over at:
https://github.com/bioconda/bioconda-recipes/pull/5894

Happy to proceed after that.

Best regards,
Sebastian

**Comment by jcupitt** [comment]
Hello, oh dear, I need to look at this again. I'll try this weekend. Sorry for being so slow.

**Comment by sebastian-luna-valero** [comment]
Great, many thanks!

No worries, it also slipped my mind. Many other things on top of the pile...

**Comment by saketkc** [comment]
HI @jcupitt and @sebastian-luna-valero, any updates on getting this into conda (-forge/bio-)?

**Comment by oeway** [comment]
Hi, same question as above, would be great to have it in conda.

**Comment by jxu** [comment]
What happened to the conda PR?

**Comment by oeway** [comment]
Hi, 

This is a gentle ping to @jcupitt and @sebastian-luna-valero. 

I am trying to distribute a web tool which relies on libvips to generate deepzoom/zoomify files, however, I am stuck because some OS cannot build libvips, and docker isn't really a solution for the end user.

Following the discussion, it seems you guys are very close to a more accessible conda package. A lot more people will benifit from if that really happened. 

Since the PR to bioconda was closed, would the next plan be submitting to conda-forge? 

Thanks a lot, your help will be highly appreciated.

**Comment by jcupitt** [comment]
I has a quick go at a `meta.yml` for pyvips:

```yaml
{% set name = "pyvips" %}
{% set version = "2.1.8" %}

package:
  name: "{{ name|lower }}"
  version: "{{ version }}"

source:
  url: https://github.com/libvips/{{ name }}/archive/v{{ version }}.tar.gz
  sha256: 4dcd1d6689f3faf46b340e70ee238bfd323b9757c5aff8d22a9e05201ebf2cc3

build:
  number: 0
  script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv "

requirements:
  host:
    - pip
    - pkgconfig
    - pytest-runner
    - python
    - cffi >=1.0
  run:
    - python
    - pkgconfig
    - cffi >=1.0

test:
  imports:
    - pyvips

about:
  home: https://github.com/libvips/pyvips
  license: MIT
  license_family: MIT
  summary: Binding for libvips
  doc_url: https://libvips.github.io/pyvips/

extra:
  recipe-maintainers:
    - jcupitt
```

It seems to work, though of course that's picking up libvips from the host rather than downloading a binary.

I'll see if I can make conda build its own libvips binary as well.

**Comment by jcupitt** [comment]
I've made a branch which seems to make a working libvips. It passes their CI tests too.

Here's the configure summary, you can see lots of useful things are missing. 

```
use fftw3 for FFT: 			yes
Magick package: 			MagickCore
Magick API version: 			magick7
load with libMagick: 			yes
save with libMagick: 			yes
accelerate loops with orc: 		no
  (requires orc-0.4.11 or later)
ICC profile support with lcms: 		no
file import with niftiio: 		no
file import with libheif: 		no
file import with OpenEXR: 		no
file import with OpenSlide:		no
  (requires openslide-3.3.0 or later)
file import with matio: 		yes
PDF import with PDFium 			no
PDF import with poppler-glib: 		yes
  (requires poppler-glib 0.16.0 or later)
SVG import with librsvg-2.0: 		yes
  (requires librsvg-2.0 2.34.0 or later)
zlib: 					yes
file import with cfitsio: 		yes
file import/export with libwebp:	yes
  (requires libwebp, libwebpmux, libwebpdemux 0.5.0 or later)
text rendering with pangoft2: 		yes
file import/export with libpng: 	yes (pkg-config libpng >= 1.2.9)
  (requires libpng-1.2.9 or later)
support 8bpp PNG quantisation:		no
  (requires libimagequant)
file import/export with libtiff:	yes (pkg-config libtiff-4)
file import/export with giflib:		yes (found by search)
file import/export with libjpeg:	yes (pkg-config)
image pyramid export:			no
  (requires libgsf-1 1.14.26 or later)
use libexif to load/save JPEG metadata: no
```

- openslide, since that's in bioconda and this is a conda-forge recipe. I suppose we'd need to move back to bioconda, or ask for openslide to move to conda-forge.
- lcms, libexif is serious, EXIF metadata and colour management is important. I guess we shuld make conda-forge packages for them too.
- libgsf-1 will be annoying, it means no deepzoom output. Again, we'd need to package it.
- niftiio is annoying, I guess this is in bioconda as well.
- libheif means no HEIC support. Again, we'd need to package that separately.
- orc means libvips will not have any vector support. Things like convolution will be up to about 3x slower.

I had to pin pango to 1.40.14 to get librsvg and imagemagick to go on together.

It passes their CI tests, ie. 

```
john@kiwi:~/GIT/conda/staged-recipes (master)$ ./.circleci/run_docker_build.sh
```

I left everyone's names on the maintainer list, I hope that's OK.

imagemagick is built with jpeg, not libjpeg-turbo, so we must use jpeg too. This is very unfortunate. It would be better if IM could move to turbo as well.

**Comment by jcupitt** [comment]
So ... action plan:

- Important missing libs: openslide, lcms, libexif, libgsf-1. We should do something about these.
- Minor missing libs: openexr, niftiio, libheif, orc
- Other: can we move imagemagick to libjpeg-turbo somehow?
- Resolve some of the above, then open a PR for conda-forge
- Once that's in, make another PR for pyvips.

**Comment by jcupitt** [comment]
I added a formula for pyvips as well. To test:

```
git clone git@github.com:jcupitt/staged-recipes.git
cd staged-recipes
git checkout  add-libvips-8.8.1
conda build recipes/libvips
```

You can also try a CI build:

```
./.circleci/run_docker_build.sh
```

Likewise for pyvips, though you'll need to comment out the libvips dep.

```
git checkout add-pyvips-2.1.8
conda build recipes/pyvips
```

**Comment by jcupitt** [comment]
I thought about it and opened a PR for the recipe. Let's continue discussion there.

**Comment by jcupitt** [comment]
@oeway don't get your hopes up, this libvips conda package will not support deepzoom output.

That needs `libgsf-1`, and conda does not have a package for it. Someone needs to package that library first.

**Comment by jcupitt** [comment]
**Final summary**: it looks like this will get into conda soon. But it's going to be slow and missing lots of basic features, so I don't know how useful it will be.

I don't plan to do any more work on it, but if someone has some free time, here's a summary of what's missing:

* `lcms` http://www.littlecms.com ... libvips uses this for colour management. It would help other recipes, like imagemagick, so it'd be a great thing for someone to package it. 
* `libexif`https://libexif.github.io ... libvips uses to manage metadata in JPG, TIF, PNG and WEBP images. Again, this is is something that needs fixing urgently.
* `libgsf-1` https://github.com/GNOME/libgsf ... this is needed for deepzoom output. No deepzoom pyramids until this gets packaged.
* `openslide` ... this is in bioconda, so we can't use it. This recipe needs to move (or be copied?) to conda-forge.
* `orc` https://gitlab.freedesktop.org/gstreamer/orc ... libvips uses this for SIMD support. Operations like convolution will be about 3x slower without it.
* `libheif` https://github.com/strukturag/libheif ... libvips uses this for HEIC support. This would be a tricky one to package, it's very complex. No HEIC support until it's done.
* `niftiio` ... used for .nii.gz support. I use this, perhaps no one else does though.
* `openexr` ... I doubt if anyone will miss this heh.
* `imagemagick` has been built with `jpeg`, not `libjpeg-turbo`, so libvips is sadly limited to `jpeg` as well. This means JPEG load and save will be about 3x slower than it should be. Someone needs to persuade conda to swap to `libjpeg-turbo`.

**Comment by oeway** [comment]
@jcupitt Thanks a lot for your efforts, and thanks for the nice summary. I don't have the right expertise to work on these, but hopefully someone can pick it up.

**Comment by sebastian-luna-valero** [comment]
Hi,

I would like to help. I am trying to build libgsf. Let's see how it goes.

Best regards,
Sebastian

xref: https://gitlab.gnome.org/GNOME/libgsf/issues/18

**Comment by jcupitt** [comment]
Hey @sebastian-luna-valero, that's great!

I think this error comes from building from git. Instead, build from one of the libgsf tarballs.

```
wget https://ftp.gnome.org/pub/GNOME/sources/libgsf/1.14/libgsf-1.14.46.tar.xz
tar xf libgsf-1.14.46.tar.xz 
cd libgsf-1.14.46/
./configure ...
```

Don't run autogen!

**Comment by jcupitt** [comment]
How this works: autotools has two main modes -- maintainer mode, and distribution mode. 

In maintainer mode, you can edit `configure.ac` and all of the files that control the build system, and in distribution mode, you can run `configure` (a huge shell script) and nothing else. `autogen.sh` is a shell script that conventionally regenerates `configure` for you.

In maintainer mode, you need to have everything installed that might possibly be needed, since it needs to be able to pull in the files that control `gtk-doc` detection. In distribution mode, everything is frozen into the build system (`configure`) and all you need to have installed is a posix shell.

tldr: if you are running `autogen.sh`, you are in maintainer mode and you must have *everything* installed. If you have a `configure` file, it should all be set up for you, and should just work.

**Comment by sebastian-luna-valero** [comment]
Thanks @jcupitt 

Here we go: https://github.com/conda-forge/staged-recipes/pull/9140

**Comment by jcupitt** [comment]
Great! Once that's accepted, I'll add it to my libvips recipie.

**Comment by jcupitt** [comment]
OK, libvips package updated for libgsf.

**Comment by sebastian-luna-valero** [comment]
Thanks!

Adding:
* lcms: https://github.com/conda-forge/staged-recipes/pull/9155
* libexif: https://github.com/conda-forge/staged-recipes/pull/9156
* https://github.com/strukturag/libheif depends on https://github.com/strukturag/libde265

I will try to work on these.

**Comment by jcupitt** [comment]
OK, I've added lcms2 and libexif to the libvips recipe, and it all seems to be linking correctly.

Nice job!

**Comment by jcupitt** [comment]
The big missing feature now is openslide support.

Do you think we should ask for openslide to move from bioconda to conda-forge? Or perhaps it can be in both?

**Comment by jcupitt** [comment]
Here's something to add openslide to conda-forge:

https://github.com/conda-forge/staged-recipes/pull/9169

And another to remove it from bioconda:

https://github.com/bioconda/bioconda-recipes/pull/16938

**Comment by jcupitt** [comment]
OK, openslide is in there!

orc is the remaining big one. Could you package it, @sebastian-luna-valero?

**Comment by sebastian-luna-valero** [comment]
Thanks!

I am adding orc here: https://github.com/conda-forge/staged-recipes/pull/9173

However, it would have to be named anything other than `orc` due to a clash with: 
https://anaconda.org/conda-forge/orc

**Comment by jcupitt** [comment]
Great! I've added gsteamer-orc, let's see if it builds. After that, we can add pyvips.

**Comment by jcupitt** [comment]
OK, gstreamer-orc is in all it all seems to work! We can merge now.

**Comment by jcupitt** [comment]
@strets123 @sebastian-luna-valero could you post a message here saying you are willing to be listed as maintainers?

https://github.com/conda-forge/staged-recipes/pull/8912#discussion_r314944245

Otherwise I'll have to take you off the maintainer list.

**Comment by jcupitt** [comment]
libvips is merged!! Well done everyone.

pyvips is next: https://github.com/conda-forge/staged-recipes/pull/8926

**Comment by sebastian-luna-valero** [comment]
Brilliant, thank you all!

**Comment by jcupitt** [comment]
And pyvips is is now. All testing very welcome.

**Comment by jcupitt** [comment]
I wrote a quick post on conda install:

https://libvips.github.io/libvips/2019/08/18/pyvips-in-conda.html

Does that look OK @sebastian-luna-valero ? I can remove your name if you prefer.

**Comment by sebastian-luna-valero** [comment]
Many thanks for the acknowledgement, it looks good to me.
